From 13b0273dd3c2035b0a147dc8b80252057142c28c Mon Sep 17 00:00:00 2001
From: Damian Kaczka <damiankaczka@outlook.com>
Date: Wed, 24 Dec 2025 16:21:50 +0100
Subject: [PATCH] avahi-daemon: fix incorrect memmove length in simple protocol
 server

Fix a potential integer underflow when less than outbuf_length
bytes are written to the file descriptor.

It's unlikely that this bug is triggerable because the output buffer
is fixed and small, which reduces the chances of a partial write.

Fixes https://github.com/avahi/avahi/issues/715
Reported-by: Joshua Rogers <joshua.hu> using the ZeroPath tool <zeropath.com>
---
 avahi-daemon/simple-protocol.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/avahi-daemon/simple-protocol.c b/avahi-daemon/simple-protocol.c
index 28be512..6397891 100644
--- a/avahi-daemon/simple-protocol.c
+++ b/avahi-daemon/simple-protocol.c
@@ -419,8 +419,7 @@ static void client_work(AvahiWatch *watch, AVAHI_GCC_UNUSED int fd, AvahiWatchEv
         assert((size_t) r <= c->outbuf_length);
         c->outbuf_length -= r;
 
-        if (c->outbuf_length)
-            memmove(c->outbuf, c->outbuf + r, c->outbuf_length - r);
+        memmove(c->outbuf, c->outbuf + r, c->outbuf_length);
 
         if (c->outbuf_length == 0 && c->state == CLIENT_DEAD) {
             client_free(c);
